text/0061-functions.md

```diff
@@ -0,0 +1,337 @@
+
+### Basic Example
+
+Looking at the linked issue, the *semantic versioning* (SemVer) use case is a perfect use-case for functions. SemVer can be trivially encoded within Cedar’s existing expression language.
+
+```
+// Permit access when the api is greater than 2.1
+permit (principal, action == Action::"someAction", resource)
+when {
+    resource.apiVersion.major > 2 ||
+    (resource.apiVersion.major == 2 && resource.apiVersion.minor >= 1)
+};
+```
+
+Here it is instead represented with functions
+
+```
+def semver(?major, ?minor, ?long) 
+  { major : ?major, minor : ?minor, patch : ?patch }
+;
+
+// lhs > rhs ?
+def semverGT(?lhs, ?rhs) 
+  if ?lhs.major == ?rhs.major then 
+    if ?lhs.minor == ?rhs.minor then
+      ?lhs.patch > ?rhs.patch
+    else
+      ?lhs.minor > ?rhs.minor
+  else
+    ?lhs.major > ?rhs.major
+;
+
+// Permit access when the api is greater than 2.1
+permit (principal, action == Action::"someAction", resource)
+when {
+     semverGT(resource.apiVersion, semver(2,1,0))
+};
+```
+
+For simplicity, safety, and readability, Cedar macros cannot call other macros, and cannot take macros as arguments.
+
+## Motivation
+
+Cedar currently lacks mechanisms for users to build abstractions. The only existing mechanism is extension functions, which are insufficient for two reasons:
+
+1. They are extremely heavyweight, requiring modifying the source of the Cedar evaluator. This means that users who want to stay on official versions of Cedar have no choice but to attempt to submit a PR and get it accepted into the mainline. This process does not scale.
+    1. For data structures that are relatively standard (ex: SemVer, or OID Users as proposed in [RFC 58](https://github.com/cedar-policy/rfcs/blob/cdisselkoen/standard-library/text/0058-standard-library.md)), it’s hard to know what’s in-demand enough to be included, and how to balance that against avoiding bloat. There’s no way to naturally observe usage because the only way to “install” the extension pre-acceptance is to vend a modified version of Cedar.
+    2. Users may have data structures that are totally bespoke to their systems. It makes no sense to include these in the standard Cedar distribution at all, yet users may still want some way to build abstractions. 
+2. They are too powerful. Extensions are implemented via arbitrary Rust code, which is essential for encoding features that cannot be represented via Cedar expressions (such as IP Addresses), but opens the door for a wide range of bugs/design issues. It’s trivial to design an extension that is difficult to validate and/or logically encode for analysis. Problematically, extension functions can potentially exhibit non-determinism, non-termination, or non-linear performance; interact with the operating system; or violate memory safety. This raises the code review burden when considering an extension function's implementation.
```

Comment:
This is a great summary of the current state of extension functions. This is why I'm in favor of this RFC: there is currently no way to build abstractions in Cedar outside of (built-in) extension functions.

So although I prefer "macro" to "function" for the feature in this RFC, I'd like to throw another name into the ring: _User-defined extension functions_

(Note that this would make the most sense if we used CBV for both types of calls.)

Reply:
> User-defined extension functions

I prefer not to call them functions because they are not CBV.